File: src/js/lib/value-agreement.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

import { AsciiCharSet } from './ascii-char-set.js';
import { Agreement, contribution } from './rank-agreement.js';
import { RxFactory } from './rx.js';
import { minterms } from './state-map.js';
import { COPY_MARKER, EotKind, TxFactory, convert } from './tx.js';

/**
 * How many product states the walk may build before giving up.
 *
 * The same figure as `MAX_PAIRS` in `rank-agreement.js`, for the same reason: a comparison worth
 * making is between two naxps that resemble one another, and none of the pairs anybody has reason
 * to compare has needed more than a few hundred.
 */
export const MAX_TUPLES = 200_000;

/**
 * Walks a concrete emission through a canonical machine, returning what it adds to the running
 * total and where it ends up.
 *
 * @param {import('./state-map.js').State} from The state to walk from.
 * @param {string} emitted The emission.
 * @returns {{added: bigint, to: import('./state-map.js').State}} The amount and the state.
 */
function walk(from, emitted) {
	let added = 0n;
	let state = from;

	for (const c of emitted) {
		if (c === COPY_MARKER) { throw new Error('An undecided copy reached the walk.'); }

		const code = c.charCodeAt(0);
		let taken = null;

		for (const transition of state.transitions) {
			if (!transition.set.isEmpty && transition.set.contains(code)) {
				taken = transition;
				break;
			}
		}

		// A partial parse's output is a prefix of a canonical string, so it cannot fall off the
		// machine of the canonical language.
		if (taken === null) {
			throw new Error('A canonical form is not accepted by its own machine.');
		}

		added += contribution(state, taken, code);
		state = taken.next;
	}

	return { added, to: state };
}

/**
 * Whether any of a derivative's moves emits a character whose identity the block has not fixed.
 *
 * @param {import('./tx.js').TxDerivative} derivative The moves.
 * @returns {boolean} Whether one copies.
 */
function copies(derivative) {
	for (const move of derivative.moves) {
		if (move.emitted.includes(COPY_MARKER)) { return true; }
	}

	return false;
}

/**
 * Explores the product states reachable on a common input, reporting the first shared string the
 * two value differently.
 */
class Product {
	/**
	 * @param {import('./compiler.js').Compilation} a The first naxp.
	 * @param {import('./compiler.js').Compilation} b The second naxp.
	 * @param {TxFactory} leftFactory The first naxp's transducer factory.
	 * @param {TxFactory} rightFactory The second naxp's transducer factory.
	 * @param {number} maxTuples The budget.
	 */
	constructor(a, b, leftFactory, rightFactory, maxTuples) {
		this.a = a;
		this.b = b;
		this.leftFactory = leftFactory;
		this.rightFactory = rightFactory;
		this.maxTuples = maxTuples;

		/** @type {Map<string, number>} */
		this.indexOf = new Map();
		/** @type {Array<{left: import('./tx.js').Tx, right: import('./tx.js').Tx,
		 *   leftState: import('./state-map.js').State, rightState: import('./state-map.js').State}>} */
		this.keys = [];
		/** @type {bigint[]} */
		this.differences = [];
		/** @type {number[]} */
		this.parents = [];
		/** @type {number[]} */
		this.arrivals = [];
		/** @type {Array<{existing: number, parent: number, arrival: number}>} */
		this.conflicts = [];
		/** @type {Map<string, string | null>} */
		this.completions = new Map();
	}

	/**
	 * @param {import('./tx.js').Tx} left The first naxp's transduction.
	 * @param {import('./tx.js').Tx} right The second naxp's transduction.
	 * @returns {{agreement: string, witness: string | null}} The verdict, and where they differ.
	 */
	run(left, right) {
		const start = this.add(
			{ left, right, leftState: this.a.canonical.start, rightState: this.b.canonical.start },
			0n,
			-1,
			0);

		const queue = [start];
		let head = 0;

		while (head < queue.length) {
			const index = queue[head++];
			const ending = this.atEndOfText(index);

			if (ending === Agreement.Differs) {
				return { agreement: Agreement.Differs, witness: this.path(index) };
			}

			if (ending === Agreement.Undecided) {
				return { agreement: Agreement.Undecided, witness: null };
			}

			const key = this.keys[index];
			const sets = [...key.left.getFirstSets(), ...key.right.getFirstSets()];

			if (sets.length === 0) { continue; }

			// `minterms` returns a fresh list, so narrowing, which re-enters the step, cannot
			// disturb the one being iterated.
			for (const block of minterms(sets)) {
				if (!this.tryStep(index, block, queue)) {
					return { agreement: Agreement.Undecided, witness: null };
				}
			}
		}

		// A differing arrival matters only where the two residuals still share a string, so that
		// the differing totals are ever compared. Where they do, the two paths through the state
		// finish at different differences, so at least one of them is a shared string the two
		// value differently; but only one need be, and which is not known without the rest of the
		// walk, so each is asked the direct way.
		for (const { existing, parent, arrival } of this.conflicts) {
			const key = this.keys[existing];
			const completion = this.completion(key.left, key.right);

			if (completion === null) { continue; }

			let first = this.path(existing) + completion;
			let second = this.path(parent) + String.fromCharCode(arrival) + completion;

			if (second.length < first.length) { [first, second] = [second, first]; }

			if (this.a.encode(first) !== this.b.encode(first)) {
				return { agreement: Agreement.Differs, witness: first };
			}

			if (this.a.encode(second) !== this.b.encode(second)) {
				return { agreement: Agreement.Differs, witness: second };
			}

			throw new Error('Two arrivals differed and neither is a witness.');
		}

		return { agreement: Agreement.Agrees, witness: null };
	}

	/**
	 * Whether both sides can end the input here with different values.
	 *
	 * @param {number} index The product state.
	 * @returns {string} One of {@link Agreement}.
	 */
	atEndOfText(index) {
		const key = this.keys[index];

		if (!key.left.isNullable || !key.right.isNullable) { return Agreement.Agrees; }

		const left = key.left.getEot();
		const right = key.right.getEot();

		if (left.kind === EotKind.TooLong || right.kind === EotKind.TooLong) {
			return Agreement.Undecided;
		}

		if (left.kind !== EotKind.Single || right.kind !== EotKind.Single) {
			throw new Error('A compiled naxp emits more than one thing at end of text.');
		}

		// Ending emits whatever the residual still owes, which the canonical machine has to be
		// walked over. The end of text transition itself contributes nothing on either side, and
		// the one that turns a total into a value cancels.
		const leftWalk = walk(key.leftState, left.text);
		const rightWalk = walk(key.rightState, right.text);

		if (!leftWalk.to.acceptsEndOfText || !rightWalk.to.acceptsEndOfText) {
			throw new Error('A canonical form is not accepted by its own machine.');
		}

		return this.differences[index] + leftWalk.added - rightWalk.added === 0n
			? Agreement.Agrees
			: Agreement.Differs;
	}

	/**
	 * Takes one step of the input, narrowing the block to single characters where either side
	 * would copy one.
	 *
	 * A copied character has to be known before it can be walked through a canonical machine,
	 * since the rank it contributes depends on which character it is. That is the only reason to
	 * narrow. The W3 checker can let two identical copies cancel because it compares them with
	 * each other; here they are walked through two different machines, so they cannot.
	 *
	 * @param {number} index The product state.
	 * @param {AsciiCharSet} block The block to step by.
	 * @param {number[]} queue The queue to append to.
	 * @returns {boolean} Whether the step was decided.
	 */
	tryStep(index, block, queue) {
		const key = this.keys[index];
		const left = this.leftFactory.derivative(key.left, block);
		const right = this.rightFactory.derivative(key.right, block);

		if (left.tooLong || right.tooLong) { return false; }

		if (left.skipsAmbiguously || right.skipsAmbiguously) {
			throw new Error('A compiled naxp skips ambiguously.');
		}

		// One side cannot consume this block, so no shared string passes through it.
		if (left.moves.length === 0 || right.moves.length === 0) { return true; }

		if (block.singleCharacter === null && (copies(left) || copies(right))) {
			for (const code of block) {
				if (!this.tryStep(index, AsciiCharSet.fromSingleChar(code), queue)) { return false; }
			}

			return true;
		}

		const arrival = block.singleCharacter ?? block.characterAt(0);
		const difference = this.differences[index];

		for (const leftMove of left.moves) {
			for (const rightMove of right.moves) {
				const leftWalk = walk(key.leftState, leftMove.emitted);
				const rightWalk = walk(key.rightState, rightMove.emitted);
				const next = difference + leftWalk.added - rightWalk.added;
				const nextKey = {
					left: leftMove.residual,
					right: rightMove.residual,
					leftState: leftWalk.to,
					rightState: rightWalk.to,
				};

				const existing = this.indexOf.get(keyOf(nextKey));

				if (existing !== undefined) {
					if (this.differences[existing] !== next) {
						this.conflicts.push({ existing, parent: index, arrival });
					}

					continue;
				}

				if (this.keys.length >= this.maxTuples) { return false; }

				queue.push(this.add(nextKey, next, index, arrival));
			}
		}

		return true;
	}

	/**
	 * @param {{left: import('./tx.js').Tx, right: import('./tx.js').Tx,
	 *   leftState: import('./state-map.js').State,
	 *   rightState: import('./state-map.js').State}} key The product state.
	 * @param {bigint} difference The first side's running total less the second's.
	 * @param {number} parent The index it was reached from.
	 * @param {number} arrival The character code that reached it.
	 * @returns {number} Its index.
	 */
	add(key, difference, parent, arrival) {
		const index = this.keys.length;

		this.indexOf.set(keyOf(key), index);
		this.keys.push(key);
		this.differences.push(difference);
		this.parents.push(parent);
		this.arrivals.push(arrival);

		return index;
	}

	/**
	 * The input that reaches a state, read back along the path that found it.
	 *
	 * @param {number} index The state.
	 * @returns {string} The input.
	 */
	path(index) {
		const codes = [];

		for (let at = index; this.parents[at] >= 0; at = this.parents[at]) {
			codes.unshift(this.arrivals[at]);
		}

		return String.fromCharCode(...codes);
	}

	/**
	 * A string both residuals accept, or null where there is none.
	 *
	 * Existence needs no narrowing: what a block emits varies by character, but whether it is
	 * consumed does not.
	 *
	 * @param {import('./tx.js').Tx} left The first residual.
	 * @param {import('./tx.js').Tx} right The second residual.
	 * @returns {string | null} The completion, or null.
	 */
	completion(left, right) {
		const key = `${left.id},${right.id}`;
		const known = this.completions.get(key);

		if (known !== undefined) { return known; }

		// Null while the answer is being worked out, so a pair cannot depend on itself.
		this.completions.set(key, null);

		let found = null;

		if (left.isNullable && right.isNullable) {
			found = '';
		} else {
			const sets = [...left.getFirstSets(), ...right.getFirstSets()];

			for (const block of minterms(sets)) {
				const leftDerivative = this.leftFactory.derivative(left, block);
				const rightDerivative = this.rightFactory.derivative(right, block);

				if (leftDerivative.moves.length === 0 || rightDerivative.moves.length === 0) {
					continue;
				}

				for (const leftMove of leftDerivative.moves) {
					for (const rightMove of rightDerivative.moves) {
						const rest = this.completion(leftMove.residual, rightMove.residual);

						if (rest !== null) {
							found = String.fromCharCode(block.characterAt(0)) + rest;
							break;
						}
					}

					if (found !== null) { break; }
				}

				if (found !== null) { break; }
			}
		}

		this.completions.set(key, found);

		return found;
	}
}

/**
 * A product state written out: where each parse has got to, and where each canonical machine has
 * got to on what that parse has emitted.
 *
 * The two transductions come from different factories, so identity is by side and id rather than
 * by reference alone; the same holds of the two machines.
 *
 * @param {{left: import('./tx.js').Tx, right: import('./tx.js').Tx,
 *   leftState: import('./state-map.js').State,
 *   rightState: import('./state-map.js').State}} key The product state.
 * @returns {string} A string equal for equal states.
 */
function keyOf(key) {
	return `${key.left.id}|${key.right.id}|${key.leftState.id}|${key.rightState.id}`;
}

/**
 * Whether `a` and `b` give the same value to every string both accept.
 *
 * This is the question the encoding relation turns on, and it is not the question of whether the
 * two canonicalise alike. `(A|B)!A` and `(A|B)!B` print different things for `B` and give it the
 * same value, because each rank is taken in its own canonical language. So canonical forms are
 * never compared here. Each side's output is fed into its own canonical machine as it is emitted,
 * and what is carried is the difference of the two running rank totals, as `rank-agreement.js`
 * carries it over two machines.
 *
 * A state is a pair of transduction residuals, one from each naxp, with the state each canonical
 * machine has reached on its side's output so far. Nothing is held back: a copied character is
 * fixed by narrowing the block to single characters before it is walked, a rendering is a fixed
 * string, and a skipped element's end of text output is a fixed string, so every emission is
 * concrete at the step that makes it. That is why there is no delay here where the W3 checker
 * needs one. The square compares two strings that arrive at different times; this compares two
 * numbers, and a number can be accumulated as its string arrives.
 *
 * The difference at a state is a property of the state, not of the path: from a state every
 * completion both sides accept has one output per side, by W3, so the rank each side adds over it
 * is fixed, and two arrivals with different differences cannot both finish at zero. A second and
 * unequal arrival is therefore a disagreement rather than a branch, with one proviso that
 * `rank-agreement.js` also makes: it counts only where the two residuals still share a completion.
 * That proviso is exercised constantly rather than rarely. The postcode against itself reaches the
 * same state by different inputs through cross pairs of parses, `\X?` taken against `\X?` skipped,
 * whose residuals want a digit and a letter next and so share nothing.
 *
 * The argument, the pairs it was tried on and the construction it replaces are in
 * `encoding/comparison-square-review.md`, with a harness in `encoding/comparison-check/`.
 *
 * @param {import('./compiler.js').Compilation} a The first naxp.
 * @param {import('./compiler.js').Compilation} b The second naxp.
 * @param {number} [maxTuples] How many product states to allow.
 * @returns {{agreement: string, witness: string | null}} The verdict, and where they differ, a
 * string both accept and value differently, the shortest the walk found.
 */
export function compare(a, b, maxTuples = MAX_TUPLES) {
	// Each side has its own factories. The sides are ordered and never compared for identity, so
	// there is nothing to gain from sharing and nothing to get wrong.
	const leftRxFactory = new RxFactory();
	const leftTxFactory = new TxFactory(leftRxFactory);
	const left = convert(a.ast, leftTxFactory, leftRxFactory);

	const rightRxFactory = new RxFactory();
	const rightTxFactory = new TxFactory(rightRxFactory);
	const right = convert(b.ast, rightTxFactory, rightRxFactory);

	return new Product(a, b, leftTxFactory, rightTxFactory, maxTuples).run(left, right);
}

File: src/cpp/src/value_agreement.cpp
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

#include "value_agreement.hpp"

#include "ascii_char_set.hpp"
#include "compiler.hpp"
#include "rank_agreement.hpp"
#include "rank_difference.hpp"
#include "rx.hpp"
#include "state_map.hpp"
#include "tx.hpp"

#include <cstddef>
#include <cstdint>
#include <deque>
#include <optional>
#include <stdexcept>
#include <string>
#include <string_view>
#include <tuple>
#include <unordered_map>
#include <utility>
#include <vector>

namespace logmu::detail::value_agreement
{
	namespace
	{
		/// A product state: where each parse has got to, and where each canonical machine has
		/// got to on what that parse has emitted.
		///
		/// The two transductions come from different factories, so identity is by side and id
		/// rather than by pointer alone; the same holds of the two machines.
		struct key
		{
			const tx* left;
			const tx* right;
			const state* left_state;
			const state* right_state;

			bool operator==(const key& other) const noexcept
			{
				return this->left->id == other.left->id
					&& this->right->id == other.right->id
					&& this->left_state->id == other.left_state->id
					&& this->right_state->id == other.right_state->id;
			}
		};

		struct key_hash
		{
			std::size_t operator()(const key& item) const noexcept
			{
				return (((((static_cast<std::size_t>(item.left->id) * 31) ^ static_cast<std::size_t>(item.right->id)) * 31) ^ static_cast<std::size_t>(item.left_state->id)) * 31) ^ static_cast<std::size_t>(item.right_state->id);
			}
		};

		struct pair_hash
		{
			std::size_t operator()(const std::pair<int, int>& item) const noexcept
			{
				return (static_cast<std::size_t>(item.first) * 31) ^ static_cast<std::size_t>(item.second);
			}
		};

		struct conflict
		{
			std::size_t existing;
			std::size_t parent;
			char arrival;
		};

		bool copies(const tx_derivative& derivative) noexcept
		{
			for (const tx_move& move : derivative.moves)
			{
				if (move.emitted.find(copy_marker) != std::string::npos)
				{
					return true;
				}
			}

			return false;
		}

		/// Walks a concrete emission through a canonical machine, returning what it adds to the
		/// running total.
		std::uint64_t walk(const state* from, std::string_view emitted, const state*& to)
		{
			std::uint64_t added = 0;
			const state* current = from;

			for (const char c : emitted)
			{
				if (c == copy_marker)
				{
					throw std::logic_error("An undecided copy reached the walk.");
				}

				const transition* taken = nullptr;

				for (const transition& arc : current->transitions)
				{
					if (!arc.set.is_empty() && arc.set.contains(c))
					{
						taken = &arc;
						break;
					}
				}

				// A partial parse's output is a prefix of a canonical string, so it cannot fall
				// off the machine of the canonical language.
				if (taken == nullptr)
				{
					throw std::logic_error("A canonical form is not accepted by its own machine.");
				}

				added += rank_agreement::contribution(*current, *taken, c);
				current = taken->next;
			}

			to = current;

			return added;
		}

		/// Explores the product states reachable on a common input, reporting the first shared
		/// string the two value differently.
		class product
		{
		public:
			product(const compilation& a, const compilation& b, tx_factory& left_factory, tx_factory& right_factory, int max_tuples)
				: a(a)
				, b(b)
				, left_factory(left_factory)
				, right_factory(right_factory)
				, max_tuples(max_tuples)
			{
			}

			agreement run(const tx* left, const tx* right, std::optional<std::string>& witness)
			{
				witness.reset();

				std::deque<std::size_t> queue;
				queue.push_back(this->add(key{left, right, this->a.canonical().start, this->b.canonical().start}, rank_difference(), std::nullopt, '\0'));

				while (!queue.empty())
				{
					const std::size_t index = queue.front();
					queue.pop_front();

					const agreement ending = this->at_end_of_text(index);

					if (ending == agreement::differs)
					{
						witness = this->path(index);

						return agreement::differs;
					}

					if (ending == agreement::undecided)
					{
						return agreement::undecided;
					}

					const key current = this->keys[index];
					this->first_sets.clear();
					const std::vector<ascii_char_set>& left_sets = current.left->get_first_sets();
					const std::vector<ascii_char_set>& right_sets = current.right->get_first_sets();
					this->first_sets.insert(this->first_sets.end(), left_sets.begin(), left_sets.end());
					this->first_sets.insert(this->first_sets.end(), right_sets.begin(), right_sets.end());

					if (this->first_sets.empty())
					{
						continue;
					}

					state_map_builder::minterms(this->first_sets, this->blocks);

					// Copied so that narrowing, which re-enters the step, cannot disturb the
					// list being iterated.
					const std::vector<ascii_char_set> snapshot = this->blocks;

					for (const ascii_char_set block : snapshot)
					{
						if (!this->try_step(index, block, queue))
						{
							return agreement::undecided;
						}
					}
				}

				// A differing arrival matters only where the two residuals still share a string,
				// so that the differing totals are ever compared. Where they do, the two paths
				// through the state finish at different differences, so at least one of them is
				// a shared string the two value differently; but only one need be, and which is
				// not known without the rest of the walk, so each is asked the direct way.
				for (const conflict& item : this->conflicts)
				{
					const key current = this->keys[item.existing];
					const std::optional<std::string> completion = this->completion(current.left, current.right);

					if (!completion.has_value())
					{
						continue;
					}

					std::string first = this->path(item.existing) + *completion;
					std::string second = this->path(item.parent) + item.arrival + *completion;

					if (second.size() < first.size())
					{
						std::swap(first, second);
					}

					if (this->a.encode(first) != this->b.encode(first))
					{
						witness = first;
					}
					else if (this->a.encode(second) != this->b.encode(second))
					{
						witness = second;
					}
					else
					{
						throw std::logic_error("Two arrivals differed and neither is a witness.");
					}

					return agreement::differs;
				}

				return agreement::agrees;
			}

		private:
			/// Whether both sides can end the input here with different values.
			agreement at_end_of_text(std::size_t index) const
			{
				const key current = this->keys[index];

				if (!current.left->is_nullable || !current.right->is_nullable)
				{
					return agreement::agrees;
				}

				const eot& left = current.left->get_eot();
				const eot& right = current.right->get_eot();

				if (left.kind == eot_kind::too_long || right.kind == eot_kind::too_long)
				{
					return agreement::undecided;
				}

				if (left.kind != eot_kind::single || right.kind != eot_kind::single)
				{
					throw std::logic_error("A compiled naxp emits more than one thing at end of text.");
				}

				// Ending emits whatever the residual still owes, which the canonical machine has
				// to be walked over. The end of text transition itself contributes nothing on
				// either side, and the one that turns a total into a value cancels.
				const state* left_end = nullptr;
				const state* right_end = nullptr;
				const std::uint64_t left_added = walk(current.left_state, left.text, left_end);
				const std::uint64_t right_added = walk(current.right_state, right.text, right_end);

				if (!left_end->accepts_end_of_text() || !right_end->accepts_end_of_text())
				{
					throw std::logic_error("A canonical form is not accepted by its own machine.");
				}

				return this->differences[index].plus(left_added).minus(right_added).is_zero()
					? agreement::agrees
					: agreement::differs;
			}

			/// Takes one step of the input, narrowing the block to single characters where
			/// either side would copy one.
			///
			/// A copied character has to be known before it can be walked through a canonical
			/// machine, since the rank it contributes depends on which character it is. That is
			/// the only reason to narrow. The W3 checker can let two identical copies cancel
			/// because it compares them with each other; here they are walked through two
			/// different machines, so they cannot.
			bool try_step(std::size_t index, ascii_char_set block, std::deque<std::size_t>& queue)
			{
				const key current = this->keys[index];
				const tx_derivative& left = this->left_factory.derivative(current.left, block);
				const tx_derivative& right = this->right_factory.derivative(current.right, block);

				if (left.too_long || right.too_long)
				{
					return false;
				}

				if (left.skips_ambiguously || right.skips_ambiguously)
				{
					throw std::logic_error("A compiled naxp skips ambiguously.");
				}

				// One side cannot consume this block, so no shared string passes through it.
				if (left.moves.empty() || right.moves.empty())
				{
					return true;
				}

				if (!block.single_character().has_value() && (copies(left) || copies(right)))
				{
					for (const char c : block)
					{
						if (!this->try_step(index, ascii_char_set::single_character(c), queue))
						{
							return false;
						}
					}

					return true;
				}

				const char arrival = block.single_character().value_or(block.character_at(0));
				const rank_difference difference = this->differences[index];

				for (const tx_move& left_move : left.moves)
				{
					for (const tx_move& right_move : right.moves)
					{
						const state* left_state = nullptr;
						const state* right_state = nullptr;
						const std::uint64_t left_added = walk(current.left_state, left_move.emitted, left_state);
						const std::uint64_t right_added = walk(current.right_state, right_move.emitted, right_state);
						const rank_difference next = difference.plus(left_added).minus(right_added);
						const key next_key{left_move.residual, right_move.residual, left_state, right_state};

						if (const auto existing = this->index_of.find(next_key); existing != this->index_of.end())
						{
							if (this->differences[existing->second] != next)
							{
								this->conflicts.push_back(conflict{existing->second, index, arrival});
							}

							continue;
						}

						if (this->keys.size() >= static_cast<std::size_t>(this->max_tuples))
						{
							return false;
						}

						queue.push_back(this->add(next_key, next, index, arrival));
					}
				}

				return true;
			}

			std::size_t add(const key& item, rank_difference difference, std::optional<std::size_t> parent, char arrival)
			{
				const std::size_t index = this->keys.size();

				this->index_of.emplace(item, index);
				this->keys.push_back(item);
				this->differences.push_back(difference);
				this->parents.push_back(parent);
				this->arrivals.push_back(arrival);

				return index;
			}

			/// The input that reaches a state, read back along the path that found it.
			std::string path(std::size_t index) const
			{
				std::string builder;

				for (std::size_t at = index; this->parents[at].has_value(); at = *this->parents[at])
				{
					builder.insert(builder.begin(), this->arrivals[at]);
				}

				return builder;
			}

			/// A string both residuals accept, or nothing where there is none.
			///
			/// Existence needs no narrowing: what a block emits varies by character, but whether
			/// it is consumed does not.
			std::optional<std::string> completion(const tx* left, const tx* right)
			{
				const std::pair<int, int> pair(left->id, right->id);

				if (const auto known = this->completions.find(pair); known != this->completions.end())
				{
					return known->second;
				}

				// Nothing while the answer is being worked out, so a pair cannot depend on
				// itself.
				this->completions.emplace(pair, std::nullopt);

				std::optional<std::string> found;

				if (left->is_nullable && right->is_nullable)
				{
					found = std::string();
				}
				else
				{
					std::vector<ascii_char_set> sets;
					const std::vector<ascii_char_set>& left_sets = left->get_first_sets();
					const std::vector<ascii_char_set>& right_sets = right->get_first_sets();
					sets.insert(sets.end(), left_sets.begin(), left_sets.end());
					sets.insert(sets.end(), right_sets.begin(), right_sets.end());

					for (const ascii_char_set block : state_map_builder::minterms(sets))
					{
						const tx_derivative& left_derivative = this->left_factory.derivative(left, block);
						const tx_derivative& right_derivative = this->right_factory.derivative(right, block);

						if (left_derivative.moves.empty() || right_derivative.moves.empty())
						{
							continue;
						}

						for (const tx_move& left_move : left_derivative.moves)
						{
							for (const tx_move& right_move : right_derivative.moves)
							{
								const std::optional<std::string> rest = this->completion(left_move.residual, right_move.residual);

								if (rest.has_value())
								{
									found = block.character_at(0) + *rest;
									break;
								}
							}

							if (found.has_value())
							{
								break;
							}
						}

						if (found.has_value())
						{
							break;
						}
					}
				}

				this->completions[pair] = found;

				return found;
			}

			const compilation& a;
			const compilation& b;
			tx_factory& left_factory;
			tx_factory& right_factory;
			int max_tuples;
			std::unordered_map<key, std::size_t, key_hash> index_of;
			std::vector<key> keys;
			std::vector<rank_difference> differences;
			std::vector<std::optional<std::size_t>> parents;
			std::vector<char> arrivals;
			std::vector<conflict> conflicts;
			std::unordered_map<std::pair<int, int>, std::optional<std::string>, pair_hash> completions;
			std::vector<ascii_char_set> blocks;
			std::vector<ascii_char_set> first_sets;
		};
	}

	agreement compare(const compilation& a, const compilation& b, std::optional<std::string>& witness, int budget)
	{
		// Each side has its own factories. The sides are ordered and never compared for
		// identity, so there is nothing to gain from sharing and nothing to get wrong.
		rx_factory left_rxs;
		tx_factory left_factory(left_rxs);
		const tx* left = tx_converter::convert(a.tree(), left_factory, left_rxs);

		rx_factory right_rxs;
		tx_factory right_factory(right_rxs);
		const tx* right = tx_converter::convert(b.tree(), right_factory, right_rxs);

		return product(a, b, left_factory, right_factory, budget).run(left, right, witness);
	}
}
